[HDR] Images inside an iframe will not be detected for HDRness unless the iframe is composited

smfr opened the pull request:
#### 1915654207cc87c4ca36fb433b5a6b41536b9982
<pre>
[HDR] Images inside an iframe will not be detected for HDRness unless the iframe is composited
<a href="https://bugs.webkit.org/show_bug.cgi?id=292947">https://bugs.webkit.org/show_bug.cgi?id=292947</a>
<a href="https://rdar.apple.com/151245454">rdar://151245454</a>

Reviewed by Said Abou-Hallawa.

When an iframe is not composited for other reasons (i.e. not scrollable, and not
site-isolated), we don&apos;t detect that there are HDR images in the content, so don&apos;t
promote layers to HDR layers. This is because the iframe document itself is not
composited, so we don&apos;t run the HDR image detection code in RenderLayerBacking.

Fix by making an iframe use content compositing layers if we detect that is
has HDR content, as we do for scrollability in
`RenderLayerCompositor::cacheAcceleratedCompositingFlagsAfterLayout()`.

To make testing easier, ensure that calling `internals.setHasHDRContentForTesting()`
on an image after it loads triggers a `imageContentChanged` call via a change
in BitmapImageSource. This will trigger the logic that reevaluates compositing.

* LayoutTests/TestExpectations: Just skip the entire directory.
* LayoutTests/compositing/hdr/iframe-with-hdr-image-expected.txt: Added.
* LayoutTests/compositing/hdr/iframe-with-hdr-image.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setHasHDRContent):
* Source/WebCore/dom/Document.h:
(WebCore::Document::setHasHDRContent): Deleted.
 Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::createContext2d): Safer-cpp fix
* Source/WebCore/platform/graphics/BitmapImage.h: Make hasHDRContentForTesting() private.
* Source/WebCore/platform/graphics/BitmapImageSource.cpp:
(WebCore::BitmapImageSource::setHasHDRContentForTesting):
* Source/WebCore/platform/graphics/BitmapImageSource.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::cacheAcceleratedCompositingFlagsAfterLayout):

Canonical link: <a href="https://commits.webkit.org/296315@main">https://commits.webkit.org/296315@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/51d28885e513b1697f02854739a0ed35c457eb2d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108114 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27777 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18197 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113324 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58612 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110077 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28472 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36327 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82087 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111062 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22561 "Found 1 new test failure: intersection-observer/intersection-observer-should-not-leak-observed-nodes.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97397 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62519 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21978 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15533 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58069 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91921 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15593 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116450 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35179 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25905 "Found 1 new test failure: fast/webgpu/nocrash/fuzz-291180.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91112 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35554 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93674 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90907 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23175 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35798 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13559 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/30994 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35080 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40635 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34814 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38172 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36481 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/platform/graphics/BitmapImageSource.cpp:
```diff
@@ -667,6 +667,17 @@ ImageOrientation BitmapImageSource::frameOrientationAtIndex(unsigned index) cons
+    bool didHaveHDRContent = hasHDRContent();
+    m_hasHDRContentForTesting = true;
+    if (hasHDRContent() == didHaveHDRContent)
+        return;
```
Can't we simplify this code a little? It is for testing and there is no harm in calling `imageContentChanged()` if the image is HDR but we also call `setHasHDRContentForTesting()`.

```
if (m_hasHDRContentForTesting)
    return;

m_hasHDRContentForTesting = true;
...
```

shallawa commented on Source/WebCore/platform/graphics/BitmapImageSource.h:
```diff
@@ -162,7 +162,7 @@ class BitmapImageSource final : public ImageSource, public CanMakeCheckedPtr<Bit
     std::optional<Color> singlePixelSolidColor() const final { return m_descriptor.singlePixelSolidColor(); }
     bool hasHDRGainMap() const final { return m_descriptor.hasHDRGainMap(); }
-    bool hasHDRContent() const final { return m_descriptor.hasHDRGainMap() || m_descriptor.hasHDRColorSpace(); }
+    bool hasHDRContent() const final { return m_descriptor.hasHDRGainMap() || m_descriptor.hasHDRColorSpace() || hasHDRContentForTesting(); }
```
Since you added `hasHDRContentForTesting()` to `BitmapImageSource::hasHDRContent()`. `BitmapImage::hasHDRContent()` does not have to check `hasHDRContentForTesting()` again.

shallawa commented on Source/WebCore/rendering/RenderLayerCompositor.cpp:
```diff
@@ -682,8 +682,20 @@ void RenderLayerCompositor::cacheAcceleratedCompositingFlagsAfterLayout()
+        return false;
+    };
+
+    bool forceCompositingMode = m_hasAcceleratedCompositing && m_renderView.settings().forceCompositingMode() && frameContentRequiresCompositing();
     if (forceCompositingMode != m_forceCompositingMode) {
```
Can't we move all the conditions to the lambda?
```
auto calculateForceCompositingMode = [&] { ... }
if (m_forceCompositingMode != calculateForceCompositingMode()) {
    m_forceCompositingMode = !m_forceCompositingMode
    ...
}
```

smfr replied:
I think it's clearer as-is.

shallawa commented on LayoutTests/compositing/hdr/iframe-with-hdr-image.html:
```diff
@@ -0,0 +1,56 @@
+<!DOCTYPE html>
```
This test has to be disabled for all platforms. It should be enabled for iOS and macOS WK2 only. And most likely you will need different expectation for iOS. Search for `compositing/hdr/hdr-basic-image.html` in the LayoutTest directory.

smfr committed:
[HDR] Images inside an iframe will not be detected for HDRness unless the iframe is composited
https://bugs.webkit.org/show_bug.cgi?id=292947
rdar://151245454

Reviewed by Said Abou-Hallawa.

When an iframe is not composited for other reasons (i.e. not scrollable, and not
site-isolated), we don't detect that there are HDR images in the content, so don't
promote layers to HDR layers. This is because the iframe document itself is not
composited, so we don't run the HDR image detection code in RenderLayerBacking.

Fix by making an iframe use content compositing layers if we detect that is
has HDR content, as we do for scrollability in
`RenderLayerCompositor::cacheAcceleratedCompositingFlagsAfterLayout()`.

To make testing easier, ensure that calling `internals.setHasHDRContentForTesting()`
on an image after it loads triggers a `imageContentChanged` call via a change
in BitmapImageSource. This will trigger the logic that reevaluates compositing.

* LayoutTests/TestExpectations: Just skip the entire directory.
* LayoutTests/compositing/hdr/iframe-with-hdr-image-expected.txt: Added.
* LayoutTests/compositing/hdr/iframe-with-hdr-image.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setHasHDRContent):
* Source/WebCore/dom/Document.h:
(WebCore::Document::setHasHDRContent): Deleted.
 Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::createContext2d): Safer-cpp fix
* Source/WebCore/platform/graphics/BitmapImage.h: Make hasHDRContentForTesting() private.
* Source/WebCore/platform/graphics/BitmapImageSource.cpp:
(WebCore::BitmapImageSource::setHasHDRContentForTesting):
* Source/WebCore/platform/graphics/BitmapImageSource.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::cacheAcceleratedCompositingFlagsAfterLayout):

Canonical link: https://commits.webkit.org/296315@main
